From 9880f0b4fa0580224e8ca35be30bd64cde793965 Mon Sep 17 00:00:00 2001
Subject: [PATCH] bridge: fix crash when set mac address of br interface

commit 9b46922e15f4d9d2aedcd320c3b7f7f54d956da7 upstream.

When I tried to set mac address of a bridge interface to a mac
address which already learned on this bridge, I got system hang.

The cause is straight forward: function br_fdb_change_mac_address
calls fdb_insert with NULL source nbp. Then an fdb lookup is
performed. If an fdb entry is found and it's local, it's OK. But
if it's not local, source is dereferenced for printk without NULL
check.

Signed-off-by: Hong Zhiguo <honkiko@gmail.com>
Signed-off-by: David S. Miller <davem@davemloft.net>

Conflicts:
    net/bridge/br_fdb.c

diff --git a/net/bridge/br_fdb.c b/net/bridge/br_fdb.c
index 9063201..91d7087 100644
--- a/net/bridge/br_fdb.c
+++ b/net/bridge/br_fdb.c
@@ -477,7 +477,7 @@ static int fdb_insert(struct net_bridge *br, struct net_bridge_port *source,
 			/* add an interface address over a learned one */
 			br_warn(br, "adding interface %s with same address "
 				"as a received packet\n",
-				source->dev->name);
+				source ? source->dev->name : br->dev->name);
 		}
 		fdb_delete(br, fdb);
 	}
